### PR TITLE
campaigns: allow env usage in campaign steps

### PR DESCRIPTION
**Pull request by LawnGnome**
This is the backend component of #15822.

### CODE REVIEW

File: doc/campaigns/references/campaign_spec_yaml_reference.md [hunk 1]
```diff
@@ -140,10 +140,60 @@ It is executed using `docker` on the machine on which the [Sourcegraph CLI (`src
...
+```
+
+### Environment array
+
+> NOTE: This feature is only available in Sourcegraph 3.23 and later.
+
+In this case, `steps.env` is an array. Each array item is either:
+
+1. An object with a single property, in which case the key is used as the environment variable name and the value the value, or
+2. A string that defines an environment variable to include from the environment `src` is being run within. This is useful to define secrets that you don't want to include in the spec file, but this makes the spec dependent on your environment and means that the campaign spec file is no longer [the sole source of truth intended by the campaigns design](../explanations/campaigns_design.md).
```
Comment (mrnugget) [inline doc/campaigns/references/campaign_spec_yaml_reference.md:170]: Should we add half a sentence to mention the impact on caching? I'm not sure, but I do like the idea of having that covered somewhere.

**Review by mrnugget — APPROVED** [review]

**Review by eseliger — APPROVED** [review]
🌟

File: doc/campaigns/references/campaign_spec_yaml_reference.md (hunk 1 above)
Comment (LawnGnome) [reply doc/campaigns/references/campaign_spec_yaml_reference.md:170]: Yep, done.

**Review by chrispine — APPROVED** [review]
Love it!

**Pull request merged**